[MNG-8561] SourceRoot should be more lenient wrt duplicates

gnodet opened the pull request:
Add includes/excludes to the SourceKey

JIRA issue: [MNG-8561](https://issues.apache.org/jira/browse/MNG-8561)

elharo commented on impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultSourceRoot.java:
```diff
@@ -136,6 +136,31 @@ public DefaultSourceRoot(final ProjectScope scope, final Language language, fina
     }
 
+    /**
+     * Creates a new instance for the given directory and scope. The language is assumed Java.
```
assumed to be

elharo commented on impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultSourceRoot.java:
```diff
@@ -136,6 +136,31 @@ public DefaultSourceRoot(final ProjectScope scope, final Language language, fina
+     */
+    public DefaultSourceRoot(
+            final ProjectScope scope,
+            final Language language,
```
If it's assumed to be Java why are we specifying it here?

elharo reviewed (changes requested):
Needs tests

desruisseaux commented:
We could also remove this `SourceKey` stuff completely. The intend was to prevent ambiguous situations or potential mistakes such as declaring the same sources for two Java releases. But those checks are superficial anyway.

gnodet commented:
I've added support for wrapping the `<resources>` and `<testResources>`, remove the `SourceKey` completely to restore previous behavior and added an IT.

elharo commented on impl/maven-core/src/main/java/org/apache/maven/project/MavenProject.java:
```diff
@@ -770,23 +738,69 @@ public Build getBuild() {
+
+            @Override
+            public int size() {
+                return (int) getEnabledSourceRoots(ProjectScope.TEST, Language.RESOURCES)
```
why the cast to int? I would expect this would return an int already. Surely it isn't big enough to require a long?

gnodet replied:
`Stream.count()` actually returns a `long` ...

gnodet replied:
Remember a stream can be generated on the fly using `Stream#generate`, so it does not need to hold all elements in memory.

elharo commented on impl/maven-core/src/main/java/org/apache/maven/project/MavenProject.java:
```diff
@@ -449,7 +417,7 @@ public Collection<SourceRoot> getSourceRoots() {
      */
     public Stream<SourceRoot> getEnabledSourceRoots(ProjectScope scope, Language language) {
-        Stream<SourceRoot> s = sources.values().stream().filter(SourceRoot::enabled);
+        Stream<SourceRoot> s = sources.stream().filter(SourceRoot::enabled);
```
Maybe s --> stream
When I see a single letter s variable, I think string, not stream.
